### ISSUE TITLE
CMake + Ninja not working?

### ISSUE THREAD

**Issue by tomoyanonymous** [nick-thompson/blueprint#116]
Hi. I found that blueprint engine does not work when building with Ninja Generator on CMake though it can be successfully compiled. ~~(Both of my project and GainPlugin Example.)~~ Update: sorry, I confirmed GainPlugin works properly.
When I inspected with a debugger,  `registerNativeRenderingHooks` function was not called on an event loop, which was normally called when built with Unix Makefiles. It finally causes an error because of the missing `window.__BlueprintNative__` variable.

```js
ReferenceError: identifier '__BlueprintNative__' undefined
  at [anon] (../Dependencies/blueprint/blueprint/duktape/src-noline/duktape.c:81729) internal
  at eval (eval:1) preventsyield
```

I could not find a specific reason but just for a bug report...

Environments:

- macOS Catalina 10.15.5
- Xcode 11.4.1
- cmake version 3.18.3(installed via homebrew)
- ninja 1.10.1(installed via homebrew)
- blueprint 5604b941e39a3e6a13633023e5b823fdd8b1269d

**Comment by tomoyanonymous** [comment]
I found a problem that `__FILE__` returns absolute path when using Unix Makefiles but **Ninja returns a relative path to project-root** like `../examples/GainPlugin/PluginProcessor.cpp`.

juce::File's constructor expects an absolute path, so maybe when used with cmake+Ninja, it seems to nice to add a preprocessor definition

```cmake
target_compile_definition(your_project PRIVATE JS_MAINFILE_PATH="path/to/main.js")
```
And in cpp files,
```diff
-   File sourceDir = File(__FILE__).getParentDirectory();
-   File bundle = sourceDir.getChildFile("jsui/build/js/main.js");
+  #ifndef JS_MAINFILE_PATH
+    #define JS_MAINFILE_PATH  File(__FILE__).getParentDirectory().getChildFile("jsui/build/js/main.js");
+  #endif
+   File bundle(JS_MAINFILE_PATH);
```

**Comment by nick-thompson** [comment]
Wow, interesting, thanks @tomoyanonymous! I wonder if we should be using something more standardized than `__FILE__`? I'd prefer to avoid preprocessor defs if we can, but might be a necessary move

**Comment by tomoyanonymous** [comment]
I found a more clear explanation.
https://github.com/approvals/ApprovalTests.cpp/issues/74#issuecomment-572207017

> - On all platforms, the Ninja program:
>   + Gives the compiler relative paths, if the build tree is inside the source tree
>   + Gives the compiler absolute paths if build tree is outside the source tree

I think no standardized way to get source file path at least language level. In C++20,  `std::source_location::current().file_name()` returns a name of source file but still not a full path.

I know it is better to avoid preprocessor definitions if we can, but `__FILE__` is also a preprocessor definition which compiler and build system define.

How about like this, it can tell that __FILE__ is not an absolute path and also no preprocessor definition for source directory given to users at compile-time.

```cpp

File source;
#ifndef PROJECT_ROOT_PATH
  static constexpr char *bar = (char *)__FILE__;
  static_assert(bar[0] == '/', "__FILE__ is not absolute path. Please specify project root path to preprocessor definition PROJECT_ROOT_PATH .");
  sourcedir  File(__FILE__).getParentDirectory();
  #else
  sourcedir = PROJECT_ROOT_PATH;
#endif
  File bundle = sourceDir.getChildFile("jsui/build/js/main.js");

```

**Comment by stfufane** [comment]
Hi, just wanted to confirm that I got the same issue on Windows 10 when building with Ninja. I used @tomoyanonymous workaround to fix it and it worked.

**Comment by nick-thompson** [comment]
Hey @stfufane @tomoyanonymous check out #197.

Rather than trying to do a preprocessor dance around the inconsitentices in __FILE__, I added a small workaround for GainPlugin and updated the docs so that new users avoid __FILE__ in the first place. Let me know what you think

**Comment by nick-thompson** [comment]
Closed with #197. Thanks all!
